src/charm.py

```diff
@@ -259,6 +261,34 @@ def _reconfigure(self, event) -> None:
                 logger.info("Deferring reconfigure: error=%r", e)
                 event.defer()
 
+        # app relations should be made aware of the new set of hosts
+        try:
+            self._update_app_relation_data()
```

Comment:
This might be worth putting inside the `with MongoDBConnection` context, so you don't have to create a new one in `_update_app_relation_data` (line 278) and you can get rid of the additional exception handling of `PyMongoError`